### PR TITLE
fix(lsp): fix some type annotations

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

**Review by clason — APPROVED** [review]

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1967,7 +1964,7 @@ end
 --  Unlike `buf_request`, this will collect all the responses from each server instead of handling them.
 --  A map of client_id:request_result will be provided to the callback
 --
----@returns (function) A function that will cancel all requests which is the same as the one returned from `buf_request`.
+---@return fun() # A function that will cancel all requests which is the same as the one returned from `buf_request`.
```
Comment (clason) [inline runtime/lua/vim/lsp.lua:1967]: The `#` here (for an anonymous return value?) doesn't translate well to vimdoc. Can we find something else?

Comment (clason) [inline runtime/lua/vim/lsp.lua:1967]:
```suggestion
---@return fun() # A function that will cancel all requests of the same type as the one returned from `buf_request`.
```

Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua]: It's to mark the remainder as comment if there is no name:

`---@return <type> [<name> [comment] | [name] #<comment>]`

I have changed it for now and named the return value `cancel` as it fits here. But maybe worth adding some extra logic to the doc generator to strip such a prefix.

Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua]: Ended up removing the "of the same type as the one returned from `buf_request`." part altogether. I don't think it adds any clarity here but rather confuses.

Comment (clason) [reply runtime/lua/vim/lsp.lua]: > But maybe worth adding some extra logic to the doc generator to strip such a prefix.

Yes, that would be best. There's a bunch of other places in the code with a similar problem.

File: runtime/lua/vim/lsp/rpc.lua [hunk 2]
```diff
@@ -319,9 +319,9 @@ end
 ---
 ---@param method (string) The invoked LSP method
 ---@param params (table|nil) Parameters for the invoked LSP method
----@param callback (function) Callback to invoke
+---@param callback fun(err: lsp.ResponseError|nil, result: any) Callback to invoke
```
Comment (clason) [inline runtime/lua/vim/lsp/rpc.lua:322]:
```suggestion
---@param callback fun(err: lsp.ResponseError | nil, result: any) Callback to invoke
```

**Pull request merged**